libs/pika/async_mpi/include/pika/async_mpi/dispatch_mpi.hpp

```diff
@@ -57,12 +57,17 @@ namespace pika::mpi::experimental::detail {
         std::decay_t<F> f;
 
 #if defined(PIKA_HAVE_STDEXEC)
-
-        using completion_signatures = pika::execution::experimental::completion_signatures<
-            pika::execution::experimental::set_value_t(MPI_Request),
-            pika::execution::experimental::set_error_t(std::exception_ptr),
-            pika::execution::experimental::set_stopped_t()>;
-
+        template <typename...>
+        using no_value_completion = completion_signatures<>;
+        using completion_signatures =
+            pika::execution::experimental::transform_completion_signatures_of<std::decay_t<Sender>,
+                pika::execution::experimental::empty_env,
+                pika::execution::experimental::completion_signatures<
+                    pika::execution::experimental::set_value_t(MPI_Request),
+                    pika::execution::experimental::set_error_t(std::exception_ptr),
+                    // this also gets added by the default SetDone template parameter
+                    pika::execution::experimental::set_stopped_t()>,
```

Comment:
The comment was mostly meant for you, not to reproduce in the code necessarily:
```suggestion
                    pika::execution::experimental::set_error_t(std::exception_ptr)>,
```